### PR TITLE
Issue 13693 - CustomImportOrderCheck

### LINKED ISSUE #13693: Use module properties macro in all templates
macros is implemented at https://github.com/checkstyle/checkstyle/issues/13422
we need to use it in all templates

this is example of update we need to make for template files - https://github.com/checkstyle/checkstyle/pull/13681/files#diff-01387dd92ef7af5a9a1f274744b44e12fbeda8302eb79072785f7d6c9d559727L40

`mvn clean test -Dtest=XdocsPagesTest,XdocsJavaDocsTest`

### PR DESCRIPTION
**Pull request by romani**
closes #13693

https://checkstyle.org/checks/imports/customimportorder.html#Properties

nuance with not easy property default value is postponed until https://github.com/checkstyle/checkstyle/issues/14178 all explanation why we need override is here.

### CODE REVIEW

**Comment by romani** [comment]
I will put annotation on method:)

**Review by rnveach — CHANGES_REQUESTED** [review]
CI is failing.

Nothing has been said why this has to be a custom override and we don't accept the default of the system we have set up, especially for an custom list of 1. I am also against adding something like this to a javadoc, as there is nothing requiring we update it if we change the default value.

https://github.com/checkstyle/checkstyle/issues/13376 doesn't say what is "broken" about our validation.

What is the current default displayed for this type, and why do we have that if we don't want it showing up?

**Comment by romani** [comment]
>Nothing has been said why this has to be a custom override

I thought https://github.com/checkstyle/checkstyle/issues/14178 explains all.

>and we don't accept the default of the system we have set up,

We need override to avoid default value charges, from "" to null. As it is pointed in issue, this property is not simple as it doesn't have class field that represents it, so we do not have easy way to get default value, as it is in very special constant. I do not know what is best way to describe default values for such cases, so to avoid long discussions I am postponing it for late to get all cases like this, as we do not know them now.

>What is the current default displayed for this type,

Empty string.

>why do we have that if we don't want it showing up ?

We do what to keep empty string. Our parents implementation is limited to get default from class field but in this case we do not have class field, so all end up to null, we do not want null.

**Comment by rnveach** [comment]
> I thought https://github.com/checkstyle/checkstyle/issues/14178 explains all.

It explains more after your update.

> all end up to null, we do not want null

This goes back to my question `why do we have that if we don't want it showing up?`. Why does all our other work arounds for our set up not fix the issue and we need to spend time on what is right now a 1 time work around?

**Comment by romani** [comment]
>why do we have that if we don't want it showing up?

Our logic is correct: get default value field. But we have nuance, there is no field here, default value is different field and applies is setter by special logic. It is what it is.  Each time I touch some check for migration, so crazy nuance pop up.  So now we know that not only for since we need so declaration in class, but also for default. Making generalization on early stages is not good, it is error prone. A way better to do automation when we see all edge cases together. @relentless-pursuit
Already try to help in it but issue is not approved, I hope draft PR will make it a bit more clear.

>Why does all our other work arounds for our set up not fix the issue

We never run into this before 

>we need to spend time on what is right now

Issue is not approved, nobody has time to investigate. I think we will catch more cases like this. I will be more clear what to do later, in more right way. I still believe we do not know enough to make automation in right way.

>which we don't know if it will ever be used again?

In this case we loose nothing. If we lucky that much, it will be simple. But I don't think so, there should be more cases like this, it sounds like pitest demands to remove absolete code

**Comment by rnveach** [comment]
I am still not following. You say our code/logic is correct, but then go on to say it isn't correct for this check. If we need an override, then either something is wrong in the code or this check doesn't follow 99% of our other rules we have set up and working.

Why can't we correct code when field can not be found to display blanks instead of `null`?
Why can't we rename field to match setter so default can be identified?
Why is `null` not valid and empty string makes a difference? And then this goes back to why aren't we correcting all `null`s like the first question?

**Comment by romani** [comment]
> or this check doesn't follow 99% of our other rules

Check is not following our naive/simple rules. It is first time we run into such implementation. 

>Why can't we correct code when field can not be found to display blanks instead of null?

it will affect bunch of other Checks that already expect null as fallback.

>Why can't we rename field to match setter so default can be identified?

it is constant, it is not a field that can set new value.

>Why is null not valid and empty string makes a difference?

because it will change default value in xdoc/html and we are not ready to change it, "null" visible for users is bad design of Checks, user can not set `null`, so we should not show what user can not set in XML config.

**Comment by romani** [comment]
I am going to finish migration this week, and start fixing issues that we postponed until end of macros migration to eliminate unfortunate cases in SiteUtils.java

**Comment by romani** [comment]
@rnveach, please consider this PR to be merged before better fix is provided.
Better to have 100% migrated modules to easily see all changes in result xml files on each update of meta model calculation.
So there will be no conversation on what is correct default value or any other values, in separate PR we will just affect versions, nothing else.

**Comment by rnveach** [comment]
> Why can't we correct code when field can not be found to display blanks instead of null?
> it will affect bunch of other Checks that already expect null as fallback.

This is just documentation. It doesn't impact check logic, so others that expect null as fallback will still have it.

> Why can't we rename field to match setter so default can be identified?
> it is constant

Constants are not properties. Properties are mutable.
https://github.com/checkstyle/checkstyle/blob/8531fec143058ddd92fab14839f2e9389ac6cf5c/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java#L361

> "null" visible for users is bad design of Checks, user can not set null, so we should not show what user can not set in XML config.

So if `null` is never a good choice, then this is why we must swap the default to blanks for this kind of situation in the macro code.

=====

I am not seeing anything to approve here when we are fixing a problem that doesn't yet exist. Everything is leading back to changing displaying `null` to just blanks in this scenario. We don't need 100% conversion to fix things that are wrong. We have done it in other PRs.

**Comment by romani** [comment]
we do not store user provided value as-is,
https://github.com/checkstyle/checkstyle/blob/8531fec143058ddd92fab14839f2e9389ac6cf5c/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java#L264C19-L264C48
final type is List, so it is conceptually null-able.

User can not set `null` even to java bean, it will be NPE. So having `null` is completely wrong and confusing value.

> So if null is never a good choice, then this is why we must swap the default to blanks for this kind of situation in the macro code.

We do have nulls already in xdoc visible for user, to avoid "breaking changes" like updates and discussions, it was easier to just keep it as before and postpone it until a time it is actually required.

> Everything is leading back to changing displaying null to just blanks in this scenario. 

some modules are good with `null` for java beans. I do not like fact that we show users `null` but rethinking on this should be separate project, it is not easy in many cases.

> I am not seeing anything to approve here when we are fixing a problem that doesn't yet exist.

we are making our meta model be aware that some Modules has design defect and it is what it is for now.


>We don't need 100% conversion to fix things that are wrong. We have done it in other PRs.

nothing is wrong with default value, it is `""`. The only problem that it is not easy to get this value, as it is part of code in setter, not a declaration.

**Comment by romani** [comment]
@rnveach, @nrmacuso, please consider this update.
it is not conflicting with update in Check that we will do, but it will make special list for "null"s for properties, I will update issue to reword it to be for fix Header Checks instead. In this case we are confident and secure that no more null will sneak to design easily.

**Comment by rnveach** [comment]
To me, everything I said above is the same but now pointed at `PROPERTY_SPECIAL_DEFAULTS`. If `null` isn't valid, then we shouldn't display it for any of the others.

**Comment by romani** [comment]
We can merge this and deal with null in separate PRs. If we can not find easy way to change Checks, hack will stay for longer, but main point that nobody can leak null anymore, as system will push to make it empty by default.

**Comment by rnveach** [comment]
The difference in a suppression is we don't have to change any code to suppress. Here we have to create a workaround, and I don't agree with it as we have come up with atleast 2-3 other solutions where we follow our current rules, make sense, and don't need an override we don't need anywhere else in 90% of our checks.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/issues/14216 is approved and will be fixed separately.
This update is just to finish macros udpates  and start dealing with https://github.com/checkstyle/checkstyle/issues/14234 .
We can do it righaway, if you approve issue.

**Comment by romani** [comment]
@nrmancuso , I am assigning this PR to you to confirm that it is not blocking with your PR of type change.

**Comment by romani** [comment]
rebased to resolve conflict.

**Review by nrmancuso — APPROVED** [review]

**Comment by romani** [comment]
@rnveach , what comment define you request for change? I see label but not sure what you do not like.
I hope current update is small step forward with gating nulls until we make breaking changes.

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/pull/14217#issuecomment-1879912059

**Comment by romani** [comment]
Yes, Main point is to migrate to macros.
And deal with Checks unfortunate design later on .

**Comment by rnveach** [comment]
So then you understand why this PR is still awaiting changes?

**Comment by romani** [comment]
@rnveach , only macros usage in this PR now.

File: config/checkstyle-non-main-files-suppressions.xml [hunk 1]
```diff
@@ -250,11 +250,6 @@
          files="src[\\/]xdocs[\\/]checks[\\/]coding[\\/]unnecessarysemicoloninenumeration.xml.template"/>
   <suppress id="propertiesMacroMustExist"
          files="src[\\/]xdocs[\\/]checks[\\/]coding[\\/]unusedlocalvariable.xml.template"/>
-  <!-- Properties macro suppressions until https://github.com/checkstyle/checkstyle/issues/13693 -->
-  <suppress id="propertiesMacroMustExist"
-         files="src[\\/]xdocs[\\/]checks[\\/]imports[\\/]customimportorder.xml.template"/>
-
-  <!-- no properties by design -->
```
Comment (rnveach) [inline config/checkstyle-non-main-files-suppressions.xml:257]: These 2 lines can't be removed unless you are saying these checks have properties now.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: config/checkstyle-non-main-files-suppressions.xml (hunk 1 above)
Comment (romani) [reply config/checkstyle-non-main-files-suppressions.xml:257]: done

Comment (romani) [reply config/checkstyle-non-main-files-suppressions.xml:257]: reverted, it is required to combine two parts of `no properties by design`, block above is same reason `no properties by design`
